File: src/main/java/victor/training/cleancode/ExtractValueObjects.java
package victor.training.cleancode;

import javax.persistence.Entity;
import javax.persistence.Id;
import java.util.List;
import java.util.stream.Collectors;

class ExtractValueObjects {

  // see tests
  public List<CarModel> filterCarModels(CarSearchCriteria criteria, List<CarModel> carModels) {
    List<CarModel> results = carModels.stream()
        .filter(carModel -> criteria.getInterval().intersects(carModel.getInterval()))
        .collect(Collectors.toList());
    System.out.println("More filtering logic ...");
    return results;
  }

  private void applyCapacityFilter() {
    System.out.println(new Interval(1000, 1600).intersects(new Interval(1250, 2000)));
  }

}

class Alta {
  private void applyCapacityFilter() {
    System.out.println(new Interval(1000, 1600).intersects(new Interval(1250, 2000)));
  }

}

class MathUtil {

}

record Interval(int start, int end) {
  public Interval {
    if (start > end) throw new IllegalArgumentException("start larger than end");

  }
  public boolean intersects(Interval other) {
    return start <= other.end && other.start <= end;
  }
}

class CarSearchCriteria { // smells like JSON ...
  private final int startYear;
  private final int endYear;
  private final String make;

  public CarSearchCriteria(int startYear, int endYear, String make) {
    this.make = make;
    if (startYear > endYear) throw new IllegalArgumentException("start larger than end");
    this.startYear = startYear;
    this.endYear = endYear;
  }

  public int getStartYear() {
    return startYear;
  }

  public int getEndYear() {
    return endYear;
  }

  public String getMake() {
    return make;
  }

  public Interval getInterval() {
    return new Interval(startYear, endYear);
  }
}

@Entity
class CarModel { // the holy Entity Model
  @Id
  private Long id;
  private String make;
  private String model;
  private Interval yearInterval;

  public CarModel(String make, String model, Interval yearInterval) {
    this.make = make;
    this.model = model;
    this.yearInterval = yearInterval;
  }

  public Interval getInterval() {
    return yearInterval;
  }

  public Long getId() {
    return id;
  }

  public Interval getYearInterval() {
    return yearInterval;
  }

  public String getMake() {
    return make;
  }

  public String getModel() {
    return model;
  }


  @Override
  public String toString() {
    return "CarModel{" +
           "make='" + make + '\'' +
           ", model='" + model + '\'' +
           '}';
  }


}


class CarModelMapper {
  public CarModelDto toDto(CarModel carModel) {
    CarModelDto dto = new CarModelDto();
    dto.make = carModel.getMake();
    dto.model = carModel.getModel();
    dto.startYear = carModel.getYearInterval().start();
    dto.endYear = carModel.getYearInterval().end();
    dto.endYear = carModel.getYearInterval().end();
    dto.endYear = carModel.getYearInterval().end();
    dto.endYear = carModel.getYearInterval().end();
    return dto;
  }

  public CarModel fromDto(CarModelDto dto) {
    return new CarModel(dto.make, dto.model, new Interval(dto.startYear, dto.endYear));
  }
}

class CarModelDto {
  public String make;
  public String model;
  public int startYear;
  public int endYear;
}

File: src/main/java/victor/training/cleancode/ExtractValueObjects.kt
package victor.training.cleancode

import javax.persistence.Embedded

class ExtractValueObjects {
    // see tests
    fun filterCarModels(criteria: CarSearchCriteria, models: List<CarModel>): List<CarModel> {
        val criteriaInterval = criteria.yearInterval
        val results = models.filter { criteriaInterval.intersects(Interval(it.startYear, it.endYear)) }
        println("More filtering logic")
        return results
    }

    private fun applyCapacityFilter() {
        println(Interval(1000, 1600).intersects(Interval(1250, 2000)))
    }
}

class Alta {
    private fun applyCapacityFilter() {
        println(Interval(1000, 1600).intersects(Interval(1250, 2000)))
    }
}

data class Interval(val start:Int, val end:Int) {
    init {
        require(start <= end) { "start larger than end" }
    }
    fun intersects(other: Interval): Boolean {
        return start <= other.end && other.start <= end
    }
}


// smells like JSON ... FROZEN. I can't/ too expensive to change
data class CarSearchCriteria(
    private val startYear: Int,
    private val endYear: Int,
    val make: String) {

    val yearInterval: Interval
        get() = Interval(startYear, endYear)
}



class CarModel(    val id: Long? = null,
                   val make: String? = null,
                   val model: String? = null,
                   @Embedded
                   val yearInterval: Interval) {

    val startYear: Int
        get() = yearInterval.start
    val endYear: Int
        get() = yearInterval.end

}

class CarModelMapper {
    fun toDto(carModel: CarModel): CarModelDto {
        val dto = CarModelDto()
        dto.make = carModel.make
        dto.model = carModel.model
        dto.startYear = carModel.startYear
        dto.endYear = carModel.endYear
        return dto
    }

    fun fromDto(dto: CarModelDto): CarModel {
        return CarModel(null, dto.make, dto.model, Interval(dto.startYear, dto.endYear))
    }
}

class CarModelDto {
    var make: String? = null
    var model: String? = null
    var startYear = 0
    var endYear = 0
}